### ISSUE TITLE
Fix for locale-dependant date formatting in imap and conch

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4937]
|<img alt="facundobatista's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| facundobatista reported|
|-|-|
|Trac ID|trac#4937|
|Type|enhancement|
|Created|2011-03-12 04:23:26Z|
|Branch|https://github.com/twisted/twisted/tree/imap-conch-date-formatting-4937|

These modules had a date formatting using time.strftime() with the '%b' formatter, that is locale dependent.

So, if you run any of these in a non-English machine, you'll break a little the protocols.

As discussed in the mail list, these dates should be formatted using always the English month abbreviation.

This patch fixes that.

Also,  I took the liberty to improve a little the code in conch/ls.py around the date formatting modifications (adding strings is really expensive).

Attachments:

* [locale-independent-formatting.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b13/b137ce8429a4e825384ccc73911e8575e87de56c/e7081bd194ef4ca47c55b7e6d9eab96bc621d339.patch) (2427 bytes) - added by facundobatista on 2011-03-12 04:24:34Z - 
* [locale-independent-formatting-with-tests.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b13/b137ce8429a4e825384ccc73911e8575e87de56c/08497a054f906d8fa09582b37b135bf8921f2257.patch) (5544 bytes) - added by facundobatista on 2011-03-14 19:50:25Z - The same patch, but with locale-modifying tests
* [locale-independent-formatting-with-tests2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b13/b137ce8429a4e825384ccc73911e8575e87de56c/c54d64acee7ee01aab5d365e6cc4493229e7bf9a.patch) (6243 bytes) - added by facundobatista on 2011-03-14 21:56:47Z - Auto-skipping tests and improved readability
* [locale-independent-formatting-with-tests3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b13/b137ce8429a4e825384ccc73911e8575e87de56c/59e51c0fa9066fa8bcc87be4b9aedec982b79490.patch) (6227 bytes) - added by facundobatista on 2011-03-15 13:09:13Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4937 4937
type__enhancement enhancement
reporter__facundobatista facundobatista
priority__normal normal
milestone__ 
branch__branches_imap_conch_date_formatting_4937 branches/imap-conch-date-formatting-4937
branch_author__cyli__facundobatista cyli, facundobatista
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1299903806000000 1299903806000000
changetime__1344869216000000 1344869216000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

Looks like this ought to be in review, since there's a patch.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_2"></a>|@cyli commented|
|-|-|

(In [31025]) Branching to 'imap-conch-date-formatting-4937'

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_3"></a>|@cyli commented|
|-|-|

(In [31026]) Apply facundobatista's patch

refs [#4937](https://github.com/twisted/twisted/issues/4937)

**Comment by twisted-trac** [comment]
|[<img alt="djfroofy's avatar" src="https://avatars.githubusercontent.com/u/89760?s=50" width="50" height="50">](https://github.com/djfroofy)<a name="note_4"></a>|@djfroofy set owner to facundobatista|
|-|-|

There is no unit test which demonstrates the bug. I think it's sufficient to write one so the test would fail on a builder set with a different locale - or maybe do something to change locale `setUp()` and set back in `tearDown()` (or with `addCleanup`) to make the test fail anywhere - though that might be a bad idea.

Some other notes:

 1. I think it helps readability to define `MONTH_NAMES` constants less trickily. I.e. just: `MONTH_NAMES = { 1: 'Jan', 2: 'Feb', ... }`

 2. `MONTH_NAMES` is defined twice - should this not be in some common place?

 3. Some parts of the patch seem unrelated to this ticket - namely in `lsLine()` you are replacing some string concat with join on list at the end; I think the latter is better but maybe should be a separate ticket/patch.

**Comment by twisted-trac** [comment]
|<img alt="facundobatista's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|facundobatista commented|
|-|-|

Hi! Thanks for the review!

There are unittests that exercise that code... the only wait to make the code fail is to change the locale, which we can do in a test (but yourself said it may not be a good idea)

Regarding your other points... 

1. usage of zip and range is not "tricky", IMO

2. I put it in different places because I couldn't find a common one (and was simple enough), if you know where I could put it, I'd be glad to move it

3. I said above, as I was touching that code, just replaced a zillion '+' for a ''.join(); the change is not that important, so I minimized overhead including it in the same patch

**Comment by twisted-trac** [comment]
|[<img alt="djfroofy's avatar" src="https://avatars.githubusercontent.com/u/89760?s=50" width="50" height="50">](https://github.com/djfroofy)<a name="note_6"></a>|@djfroofy commented|
|-|-|

Do the unit tests which exercise the code result in test failure(s)?

I'm also of the opinion also that there should be a specific test for a bug rather than relying on side-effects of other tests.

**Comment by twisted-trac** [comment]
|<img alt="facundobatista's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|facundobatista commented|
|-|-|

There I attached a new patch, this time including tests that actually exercises the same code, but changing the locale first.

**Comment by twisted-trac** [comment]
|[<img alt="djfroofy's avatar" src="https://avatars.githubusercontent.com/u/89760?s=50" width="50" height="50">](https://github.com/djfroofy)|@djfroofy set owner to @djfroofy<br>@djfroofy set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="djfroofy's avatar" src="https://avatars.githubusercontent.com/u/89760?s=50" width="50" height="50">](https://github.com/djfroofy)<a name="note_10"></a>|@djfroofy set owner to facundobatista<br>@djfroofy set status to `new`|
|-|-|

First, the new tests ERROR on my machine because of failure setting locale:

```
locale.Error: unsupported locale setting
```

I don't know of the best way to get around this. Maybe pick a more common locale which is less likely to cause failure. Then I think the test needs to be skipped instead of causing error if setting the more common locale still causes an error.

Second, for readability, in `lsLine()` on line 51 there is a mixture or vars and inline calls to get values (`str(s.st_nlink).rjust(5)` vs. `un.ljust(9)`). Maybe just make consistent by getting rid of the variables and just putting everything inline when building the list.

**Comment by twisted-trac** [comment]
|<img alt="facundobatista's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|facundobatista removed owner|
|-|-|

djfroofy: As we discussed, configured the test to being skipped if the locale is not installed (even if not running, is a good documentation to have).

Regarding the readability in lsLine(), I improved it a lot (IMHO, hope you like it).

**Comment by twisted-trac** [comment]
|[<img alt="djfroofy's avatar" src="https://avatars.githubusercontent.com/u/89760?s=50" width="50" height="50">](https://github.com/djfroofy)|@djfroofy set owner to @djfroofy<br>@djfroofy set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="djfroofy's avatar" src="https://avatars.githubusercontent.com/u/89760?s=50" width="50" height="50">](https://github.com/djfroofy)<a name="note_13"></a>|@djfroofy set owner to @habnabit<br>@djfroofy set status to `new`|
|-|-|

+1

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_15"></a>|@habnabit set status to `closed`|
|-|-|

(In [31080]) Merge imap-conch-date-formatting-4937: fixing locale-dependent date formatting

Authors: cyli, facundobatista
Reviewer: djfroofy
Fixes: [#4937](https://github.com/twisted/twisted/issues/4937)

This removes a dependency on locale-specific formatting directives used in
t.conch and t.m.imap4 and instead only uses English month abbreviations.

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)|@habnabit set status to `reopened`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_17"></a>|@habnabit commented|
|-|-|

[build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/imap-conch-date-formatting-4937) <- committed a fix to work on 2.4; buildbot is still chugging along, but it doesn't break on 2.4 anymore.

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)|@habnabit removed owner<br>@habnabit set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun set owner to @habnabit<br>@exarkun set status to `new`|
|-|-|

Thanks!  Some simple things:

  1. `testFetchInternalDateLocaleIndependent` should be `test_fetchInte...`
  1. Don't say _should be_ in a test doc string!  Say _is_!
  1. All of the `current_locale` variables should be `currentLocale` (3 or so of them in 2 different files)
  1. The indentation of `MONTH_NAMES` is kind of messed up in `imap4.py`.  Also please make it private. :)
  1. `assertEquals` is preferred over `assertEqual` (seen in `test_cftp.py`)
  1. Likewise for `MONTH_NAMES` in `ls.py`
  1. `lsLine` needs a docstring, too.

**Comment by twisted-trac** [comment]
|<img alt="facundobatista's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_21"></a>|facundobatista removed owner|
|-|-|

Attached "locale-independent-formatting-with-tests3.patch"

Addressed:

  1. Fixed.
  2. Fixed.
  3. Fixed.
  4. Made it private. Didn't understand about indentation issue.
  5. I prefer to user assertEqual over assertEquals, as the later is not documented anywhere (not in twisted docs, neither in Python docs), so it can go away at any time.
  6. Same as (4)
  7. Fixed.

**Comment by twisted-trac** [comment]
|<img alt="facundobatista's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_22"></a>|facundobatista commented|
|-|-|

One further comment regarding point 5: assertEquals is *deprecated* in Python 3 (another reason to stop using it).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

> One further comment regarding point 5: assertEquals is *deprecated* in Python 3 (another reason to stop using it).

Yes!  We are exceptionally bad at predicting future behavior of python-dev.  Some short while before python-dev decided to deprecate `assertEquals`, Twisted standardized on it.  Trial was already providing its own implementation, so we don't see any deprecation warnings from the stdlib version either.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun commented|
|-|-|

(In [31119]) Apply locale-independent-formatting-with-tests3.patch

refs [#4937](https://github.com/twisted/twisted/issues/4937)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_25"></a>|@exarkun commented|
|-|-|

Cool!  Thanks again for the patch.  I'm going to make a few minor tweaks:

  1. The `_MONTH_NAMES` indentation I mentioned was just aligning the 2nd line of args with the ( on the 1st line.
  1. `assertEqual` changed to `assertEquals` (for now :) as mentioned in my earlier comment and as discussed IRL
  1. `test_fetch...` instead of `test_Fetch...`

... test it on buildbot, and then merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_26"></a>|@exarkun commented|
|-|-|

(In [31123]) Minor tweaks to whitespace and naming, plus news fragments

refs [#4937](https://github.com/twisted/twisted/issues/4937)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_27"></a>|@exarkun commented|
|-|-|

(In [31126]) Use double try instead of try/except/finally

refs [#4937](https://github.com/twisted/twisted/issues/4937)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_28"></a>|@exarkun set status to `closed`|
|-|-|

(In [31137]) Merge imap-conch-date-formatting-4937

Author: facundobatista, habnabit
Reviewer: djfroofy, exarkun
Fixes: [#4937](https://github.com/twisted/twisted/issues/4937)

This removes a dependency on locale-specific formatting directives used in
t.conch and t.m.imap4 and instead only uses English month abbreviations.  This
re-merge fixes a Python 2.4 incompatibility.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_29"></a>|@exarkun commented|
|-|-|

This was a duplicate of [#3065](https://github.com/twisted/twisted/issues/3065).
